lsp/nls/src/requests/completion.rs

```diff
@@ -21,42 +104,149 @@ pub fn handle_completion(
+                    match i.kind {
+                        // Get record fields from static type info
+                        TermKind::Declaration(..)
+                            if name_id == i.id && matches!(ty, Types(AbsType::Record(..))) =>
+                        {
+                            match &ty {
+                                Types(AbsType::Record(row)) => {
+                                    Some((extract_ident(row), i.ty.clone()))
+                                }
+                                _ => unreachable!(),
+                            }
+                        }
+                        // Get record fields from contract metadata
+                        TermKind::Declaration(_, _, ValueState::Known(body_id))
+                            if name_id == i.id
+                                && find_record_contract(&linearization.linearization, body_id)
+                                    .is_some() =>
+                        {
+                            // unwrapping is safe here because of the pattern condition.
+                            let contract =
+                                find_record_contract(&linearization.linearization, body_id)
+                                    .unwrap();
+                            let fields = match &contract.types {
+                                Types(AbsType::Record(row)) => extract_ident(row),
+                                _ => Vec::with_capacity(0),
+                            };
+                            Some((fields, i.ty.clone()))
+                        }
+
+                        // Get record fields from lexical scoping
+                        TermKind::Declaration(_, _, ValueState::Known(body_id))
+                            if name_id == i.id =>
+                        {
+                            find_record_fields(&linearization.linearization, body_id)
+                                .map(|idents| Some((idents, i.ty.clone())))
+                                .unwrap_or(None)
+                        }
+
+                        _ => None,
+                    }
+                })
+                .collect()
+        }
+
+        // variable name completion
+        Some(..) | None => linearization
+            .get_in_scope(&item)
+            .iter()
+            .filter_map(|i| match i.kind {
+                TermKind::Declaration(ref ident, _, _) => Some((vec![ident.clone()], i.ty.clone())),
+                _ => None,
+            })
+            .collect::<Vec<_>>(),
+    };
+
+    // O(n^2)
+    fn remove_duplicates(items: &Vec<CompletionItem>) -> Vec<CompletionItem> {
+        let mut seen: Vec<CompletionItem> = Vec::with_capacity(0);
+        for item in items {
+            if seen
+                .iter()
+                .find(|seen_item| seen_item.label == item.label)
+                .is_none()
+            {
+                seen.push(item.clone())
+            }
+        }
+
+        seen
+    }
```

Comment:
Suggestions:
* use `.any()` rather than `.find().is_none()`
* use `.fold()` rather than a `for` loop

This should give something like
```rust
    fn remove_duplicates(items: &Vec<CompletionItem>) -> Vec<CompletionItem> {
        items.iter().fold(Vec::with_capacity(0), |mut seen, item| {
            if !seen.iter().any(|seen_item| seen_item.label == item.label) {
                seen.push(item.clone())
            }
            seen
        })
    }
```

Reply:
I don't think I agree with the `fold` one. Returning the `seen` at the end feels artificial. It's common when filling a container to declare a local mutable variable and fill it using a `for` loop (and there might be a performance difference, but that's harder to assess). I did the fold thing initially too, coming from langauges without a `for`, but after some time I think a `for` is actually more meaningful and readable if you're mutating the accumulator in place.